### PR TITLE
rework of Linux build scripts

### PR DESCRIPTION
**Pull request by jamincollins**
* change to ninja build system
* distribution specific logic broken out
* n/9 progress entries make no sense, removing
* {}'ing all variables
* sorting help output
* sorting package lists
* project doesn't use submodules, removing logic

### CODE REVIEW

**Comment by buzzhuzz** [comment]
It would be great to have an option to create meta package with build deps. Even meta package description file would be sufficient.

For example, on DEB systems we usually use equivs-build for this.
Contents of the Input meta package description file for equivs-build may look like this:

```
Section: misc
Priority: optional
Standards-Version: 3.9.2

Package: orcaslicer-builddeps
Version: 1
Depends:
    autoconf,
    build-essential,
    cmake,
    eglexternalplatform-dev,
    extra-cmake-modules,
    file,
    gettext,
    git,
    libcurl4-openssl-dev,
    libdbus-1-dev,
    libglew-dev,
    libgstreamerd-3-dev,
    libmspack-dev,
    libosmesa6-dev,
    libsecret-1-dev,
    libssl-dev,
    libtool,
    libudev-dev,
    libwebkit2gtk-4.0-dev,
    ninja-build,
    texinfo,
    wget

Description: Meta package for OrcaSlicer build deps
 Install all needed build dependencies for OrcaSlicer
```

After that you call `equivs-build package-description.txt` and get nice `orcaslicer-builddeps_1_all.deb`

**Comment by jamincollins** [comment]
@buzzhuzz I'm not sure what you're suggesting here?  Do you want some sort of pseudo or deps package definition for the various distributions?

If so, 
1. I don't know if all distributions support that
2. seems redundant with the `-u` option already doing the package installation

**Comment by buzzhuzz** [comment]
Basically, yes - definitions of meta packages for each distribution. But this is a nice to have feature as it is not hard to build meta package manually since your change has nice list of packages for each supported distribution.

Installation of all the packages manually using '-u' switch makes it harder in the future to track why this particular dev package has been installed and what build deps could be deleted once we stopped working on the project.

Anyways, this could be done as a separate change.

**Comment by SoftFever** [comment]
Does it still work for the Ubuntu CI/Cd pipeline?
.github\workflows\build_orca.yml

**Comment by jamincollins** [comment]
I've tested it locally with [act](https://github.com/nektos/act).  I also have some workflow changes that skip the uploads and token bits when run locally through **act**.

```
[Build OrcaSlicer/build_orca.yml/Build OrcaSlicer]   ✅  Success - Main Build slicer
[Build OrcaSlicer/build_orca.yml/Build OrcaSlicer] ⭐ Run Post load cached deps
[Build OrcaSlicer/build_orca.yml/Build OrcaSlicer]   🐳  docker exec cmd=[node /var/run/act/actions/actions-cache@v4/dist/save/index.js] user= workdir=
[Build OrcaSlicer/build_orca.yml/Build OrcaSlicer]   💬  ::debug::Cache state/key: linux-cache-orcaslicer_deps-build-6fd40b94f87c17bf20ab1131f2f8d8631805175715c84c26f6115f233c295b76
| Cache hit occurred on the primary key Linux-cache-orcaslicer_deps-build-6fd40b94f87c17bf20ab1131f2f8d8631805175715c84c26f6115f233c295b76, not saving cache.
[Build OrcaSlicer/build_orca.yml/Build OrcaSlicer]   ✅  Success - Post load cached deps
[Build OrcaSlicer/build_orca.yml/Build OrcaSlicer] Cleaning up container for job Build OrcaSlicer
[Build OrcaSlicer/build_orca.yml/Build OrcaSlicer] 🏁  Job succeeded
```

**Comment by SoftFever** [comment]
> > Does it still work for the Ubuntu CI/Cd pipeline?
> > .github\workflows\build_orca.yml
> 
> Yes, I've tested it locally with [act](https://github.com/nektos/act). I also have some workflow changes that skip the uploads and token bits when run locally through **act**.
> 
> ```
> [Build OrcaSlicer/build_orca.yml/Build OrcaSlicer]   ✅  Success - Main Build slicer
> [Build OrcaSlicer/build_orca.yml/Build OrcaSlicer] ⭐ Run Post load cached deps
> [Build OrcaSlicer/build_orca.yml/Build OrcaSlicer]   🐳  docker exec cmd=[node /var/run/act/actions/actions-cache@v4/dist/save/index.js] user= workdir=
> [Build OrcaSlicer/build_orca.yml/Build OrcaSlicer]   💬  ::debug::Cache state/key: linux-cache-orcaslicer_deps-build-6fd40b94f87c17bf20ab1131f2f8d8631805175715c84c26f6115f233c295b76
> | Cache hit occurred on the primary key Linux-cache-orcaslicer_deps-build-6fd40b94f87c17bf20ab1131f2f8d8631805175715c84c26f6115f233c295b76, not saving cache.
> [Build OrcaSlicer/build_orca.yml/Build OrcaSlicer]   ✅  Success - Post load cached deps
> [Build OrcaSlicer/build_orca.yml/Build OrcaSlicer] Cleaning up container for job Build OrcaSlicer
> [Build OrcaSlicer/build_orca.yml/Build OrcaSlicer] 🏁  Job succeeded
> ```

Awesome, thanks

File: BuildLinux.sh [hunk 1]
```diff
@@ -1,10 +1,8 @@
 #!/bin/bash
 
-export ROOT=`pwd`
-export NCORES=`nproc --all`
+export ROOT=$(dirname $(readlink -f ${0}))
+export NCORES=$(nproc --all)
 export CMAKE_BUILD_PARALLEL_LEVEL=${NCORES}
```
Comment (Ocraftyone) [inline BuildLinux.sh:5]: the parallel level is not required with ninja bc it always scales to the maximum parallelization

File: BuildLinux.sh [hunk 2]
```diff
@@ -21,238 +19,172 @@ function check_available_memory_and_disk() {
...
 
 unset name
-while getopts ":dsiuhgbr" opt; do
+while getopts ":1bcdghirsu" opt; do
   case ${opt} in
-    u )
-        UPDATE_LIB="1"
+    1 )
+        NCORES=1
+        CMAKE_BUILD_PARALLEL_LEVEL=${NCORES}
```
Comment (Ocraftyone) [inline BuildLinux.sh:49]: just do export here and get rid of NCORES

File: BuildLinux.sh [hunk 3]
```diff
@@ -21,238 +19,172 @@ function check_available_memory_and_disk() {
...
-        # popd
-        # echo "done"
 
-        # FIXME: only clean deps if compiling succeeds; otherwise reruns waste tonnes of time!
-        # clean deps
-        # echo "[6/9] Cleaning dependencies..."
-        # rm -rf dep_*
-    popd
-    echo "done"
+    echo "cmake .. ${BUILD_ARGS}"
```
Comment (Ocraftyone) [inline BuildLinux.sh:138]: update echo to show correct command

File: BuildLinux.sh [hunk 4]
```diff
@@ -21,238 +19,172 @@ function check_available_memory_and_disk() {
...
-echo "FOUND_GTK2=$FOUND_GTK2)"
-if [[ -z "$FOUND_GTK2_DEV" ]]
-then
-if [[ -z "$FOUND_GTK3_DEV" ]]
+echo "FOUND_GTK3=${FOUND_GTK3}"
+if [[ -z "${FOUND_GTK3_DEV}" ]]
 then
     echo "Error, you must install the dependencies before."
     echo "Use option -u with sudo"
     exit 0
```
Comment (Ocraftyone) [inline BuildLinux.sh:97]:
```suggestion
    exit 1
```

File: BuildLinux.sh [hunk 5]
```diff
@@ -21,238 +19,172 @@ function check_available_memory_and_disk() {
...
 fi
 
-if [[ -n "$BUILD_DEPS" ]]
+if [[ -n "${BUILD_DEPS}" ]]
 then
-    echo "[3/9] Configuring dependencies..."
+    echo "Configuring dependencies..."
     BUILD_ARGS=""
-    if [[ -n "$FOUND_GTK3_DEV" ]]
+    if [[ -n "${FOUND_GTK3_DEV}" ]]
```
Comment (Ocraftyone) [inline BuildLinux.sh:117]: remove this if statement since check is done above and it fails the program if not found

File: BuildLinux.sh (hunk 4 above)
Comment (SoftFever) [reply BuildLinux.sh:97]: Nice catch!

File: BuildLinux.sh (hunk 1 above)
Comment (jamincollins) [reply BuildLinux.sh:5]: It is needed for advanced caching

**Comment by jamincollins** [comment]
Testing one more adjustment

**Comment by jamincollins** [comment]
Should be good to go now.

**Review by Ocraftyone — APPROVED** [review]
Looks good to me

**Review by SoftFever — APPROVED** [review]
Looks good
Thank you!

**Pull request merged**